Deployment on Kubernetes support with LO4K8s

mitkus opened the pull request:
This is mostly based on https://github.com/longshorej/lagom-java-chirper-tooling-example

Issue: https://github.com/lightbend/platform-team/issues/29

mitkus committed:
Enable LO4K8s

lightbend-cla-validator commented:
Hi @mitkus,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

mitkus commented:
Just to be clear, I have signed http://www.lightbend.com/contribute/cla since validator complained about that. I think the message will not go away though.

ignasi35 commented:
close/openning the issue to force validation to rerun

ignasi35 commented on KUBERNETES.md:
```diff
@@ -0,0 +1,241 @@
+  --generate-pod-controllers --generate-services \
+  --env JAVA_OPTS="-Dplay.http.secret.key=$chirp_secret -Dplay.filters.hosts.allowed.0=$allowed_host" \
+  --external-service "cas_native=$cassandra_svc" \
+  --service-type NodePort \
```
Just curious, what's special about `chirp-impl` so that it's the only service that uses `--service-type NodePort`. I'm a kubernetes noob and barely read 5 minutes about `NodePort`. Is that required for Akka Cluster? If that's the case I think all services using persistence (cassandra) need it.

longshorej replied:
This was an oversight from my example - it should be removed. Thanks for asking @ignasi35.

ignasi35 reviewed (commented):
(oh, I had this pending review for few days, forgot to comment)

longshorej commented on KUBERNETES.md:
```diff
@@ -0,0 +1,241 @@
+chirp_secret="youmustchangeme"
+friend_secret="youmustchangeme"
+activity_stream_secret="youmustchangeme"
+front_end_secret="youmustchangeme"
```
We should make these `changeme` - see the discussion here: https://github.com/lagom/online-auction-scala/pull/62/files/fa14ccdca11a3a3a1ab49590f80463b190d747de#r169196711

longshorej commented on KUBERNETES.md:
```diff
@@ -0,0 +1,241 @@
+  "$registry/chirp-impl:1.0.0-SNAPSHOT" \
+  "$registry/friend-impl:1.0.0-SNAPSHOT" \
+  "$registry/activity-stream-impl:1.0.0-SNAPSHOT" \
+  "$registry/front-end:1.0.0-SNAPSHOT" | kubectl apply -f -
```
`$registry/` needs to be removed from all of these.

longshorej commented on project/plugins.sbt:
```diff
@@ -1,3 +1,4 @@
 addSbtPlugin("com.lightbend.lagom"     % "lagom-sbt-plugin"  % "1.4.0")
 addSbtPlugin("com.github.stonexx.sbt"  % "sbt-webpack"       % "1.2.0")
 addSbtPlugin("com.typesafe.sbteclipse" % "sbteclipse-plugin" % "5.2.4")
+addSbtPlugin("com.lightbend.rp"        % "sbt-reactive-app"  % "0.6.0")
```
`0.6.1`

longshorej commented on KUBERNETES.md:
```diff
@@ -0,0 +1,241 @@
+
+# deploy ingress
+rp generate-kubernetes-resources \
+  --ingress-path-suffix '*' \
```
This line (`--ingress-path-suffix ...`) can be removed - it was from a test on one of the cloud providers. I think our commands here should focus on Minikube.

longshorej reviewed (commented):
Thanks @mitkus, I left a few comments.

ignasi35 commented on KUBERNETES.md:
````diff
@@ -0,0 +1,241 @@
+Verify that Helm is available (this takes a minute or two):
+
+```bash
+kubectl --namespace kube-system get deploy/tiller-deploy
````
Having just gone over I think there's small differences that we could try to remove. E.g. using `-w` (see https://github.com/lagom/online-auction-scala/pull/62/files#diff-31d70497fa18936de68ddfe0d7c4c22aR89)

ignasi35 commented on KUBERNETES.md:
````diff
@@ -0,0 +1,241 @@
+Verify that it is available (this takes a minute or two):
+
+```bash
+kubectl get deploy/reactive-sandbox
````
Same as above. Could we user `-w`?

ignasi35 commented on KUBERNETES.md:
````diff
@@ -0,0 +1,241 @@
+Install the sandbox. Since Chirper only uses Cassandra, we're disabling the other services but you can leave them enabled by omitting the `set` flag if you wish.
+
+```bash
+helm install lightbend-helm-charts/reactive-sandbox --name reactive-sandbox --set elasticsearch.enabled=false,kafka.enabled=false,zookeeper.enabled=false
````
This is hard to copy as it's a long line you loose track of. How about:

```
helm install lightbend-helm-charts/reactive-sandbox \
    --name reactive-sandbox \
    --set elasticsearch.enabled=false,kafka.enabled=false,zookeeper.enabled=false
```

ignasi35 commented on KUBERNETES.md:
```diff
@@ -0,0 +1,241 @@
+
+rp generate-kubernetes-resources "activity-stream-impl:1.0.0-SNAPSHOT" \
+  --generate-pod-controllers --generate-services \
+  --env JAVA_OPTS="-Dplay.http.secret.key=$activity_stream_secret -Dplay.filters.hosts.allowed.0=$allowed_host" | kubectl apply -f -
```
Hmm, just got an error here:

```
$ rp generate-kubernetes-resources "activity-stream-impl:1.0.0-SNAPSHOT" \
→   --generate-pod-controllers --generate-services \
→   --env JAVA_OPTS="-Dplay.http.secret.key=$activity_stream_secret -Dplay.filters.hosts.allowed.0=$allowed_host" | kubectl apply -f -
[error, com.lightbend.rp.reactivecli.Main$] Akka Cluster Bootstrapping is enabled so you must specify `--pod-controller-replicas 2` (or greater), or provide `--akka-cluster-join-existing` to only join already formed clusters
error: no objects passed to apply
```

`--pod-controller-replicas 2` is missing from the command. I added that and it now worked:

```
$ rp generate-kubernetes-resources "activity-stream-impl:1.0.0-SNAPSHOT"   --generate-pod-controllers --generate-services   --env JAVA_OPTS="-Dplay.http.secret.key=$activity_stream_secret -Dplay.filters.hosts.allowed.0=$allowed_host"  --pod-controller-replicas 2     | kubectl apply -f -
deployment "activityservice-v1-0-0-snapshot" created
service "activityservice" created
```

ignasi35 replied:
Hmmm, I think this an error in the code actually. I'm not sure `activity-stream` needs clustering.

ignasi35 commented:
After making some small changes as already mentioned in some comments I ended up with:

```

minikube delete


minikube addons enable ingress
minikube start --memory 6000
eval $(minikube docker-env)

helm init
helm repo add lightbend-helm-charts https://lightbend.github.io/helm-charts
helm update

kubectl --namespace kube-system get -w deploy/tiller-deploy

helm install lightbend-helm-charts/reactive-sandbox --name reactive-sandbox --set elasticsearch.enabled=false,kafka.enabled=false,zookeeper.enabled=false

kubectl get -w deploy/reactive-sandbox

sbt clean docker:publishLocal

docker images


chirp_secret="1234"
friend_secret="1234"
activity_stream_secret="1234"
front_end_secret="1234"

cassandra_svc="_cql._tcp.reactive-sandbox-cassandra.default.svc.cluster.local"

allowed_host="$(minikube ip)"

rp generate-kubernetes-resources "chirp-impl:1.0.0-SNAPSHOT" \
  --generate-pod-controllers --generate-services \
  --env JAVA_OPTS="-Dplay.http.secret.key=$chirp_secret -Dplay.filters.hosts.allowed.0=$allowed_host" \
  --external-service "cas_native=$cassandra_svc" \
  --pod-controller-replicas 2 | kubectl apply -f -

rp generate-kubernetes-resources "friend-impl:1.0.0-SNAPSHOT" \
  --generate-pod-controllers --generate-services \
  --env JAVA_OPTS="-Dplay.http.secret.key=$friend_secret -Dplay.filters.hosts.allowed.0=$allowed_host" \
  --external-service "cas_native=$cassandra_svc" \
  --pod-controller-replicas 2 | kubectl apply -f -

rp generate-kubernetes-resources "activity-stream-impl:1.0.0-SNAPSHOT" \
  --generate-pod-controllers --generate-services \
  --env JAVA_OPTS="-Dplay.http.secret.key=$activity_stream_secret -Dplay.filters.hosts.allowed.0=$allowed_host" \
  --pod-controller-replicas 2 | kubectl apply -f -

rp generate-kubernetes-resources "front-end:1.0.0-SNAPSHOT" \
  --generate-pod-controllers --generate-services \
  --env JAVA_OPTS="-Dplay.http.secret.key=$front_end_secret -Dplay.filters.hosts.allowed.0=$allowed_host" | kubectl apply -f -

rp generate-kubernetes-resources \
  --generate-ingress --ingress-name chirper \
  "chirp-impl:1.0.0-SNAPSHOT" \
  "friend-impl:1.0.0-SNAPSHOT" \
  "activity-stream-impl:1.0.0-SNAPSHOT" \
  "front-end:1.0.0-SNAPSHOT" | kubectl apply -f -

kubectl get all


open "http://$(minikube ip)"

``` 

The end result is an open browser with an NGinx  `502`error. I waited a few minutes and eventually I saw the `fronted` UI. On that UI I get `503` for both the `login` and the `signup` form submissions. Both operations are requests to `/api/users/*`.

longshorej commented:
Hrm - seems to be working for me, both that script and `sbt 'deploy minikube'`. What does `kubectl get all` say? And if any pods aren't ready, output of `kubectl logs <pod>` would be good.

ignasi35 commented:
@longshorej can't reproduce anymore. I guess it was the dreaded "Friday Afternoon bug". 💆🏼‍♂️

mitkus committed:
Address PR review comments

mitkus commented:
Addressed the comments and made another test run to see if it works, seems ok!

Not sure about `--pod-controller-replicas 2` in deploying activity stream. For now, I've added the flag to prevent error, but looking into why it's needed.

longshorej committed:
Fix leftover text

longshorej reviewed (approved):
Thanks @mitkus! This LGTM.

(note: I've pushed a small commit to fix a leftover reference to online auction)
